File: web-client/src/services/TodoService.js
import { retryableAsync } from '../utils/observableUtils';

export default class TodoService {
  constructor(serviceUrl) {
    this._serviceUrl = serviceUrl;
  }

  async getAll() {
    try {
      const res = await retryableAsync(() => fetch(this._serviceUrl));

      if (res.status === 200) {
        return await res.json();
      }

      return Promise.reject(`${res.status}: ${res.statusText}`);
    } catch (error) {
      return Promise.reject(`Fetch Error: ${error}`);
    }
  }

  async add(item) {
    try {
      const res = await retryableAsync(() =>
        fetch(this._serviceUrl, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(item),
        })
      );

      if (res.status === 201) {
        return await res.json();
      }

      return Promise.reject(`${res.status}: ${res.statusText}`);
    } catch (error) {
      return Promise.reject(`Fetch Error: ${error}`);
    }
  }

  async update(id, item) {
    try {
      const res = await retryableAsync(() =>
        fetch(`${this._serviceUrl}/${id}`, {
          method: 'PUT',
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(item),
        })
      );

      if (res.status === 204) {
        return Promise.resolve();
      }

      return Promise.reject(`${res.status}: ${res.statusText}`);
    } catch (error) {
      return Promise.reject(`Fetch Error: ${error}`);
    }
  }

  async updateCompleted(id, completed) {
    try {
      const res = await retryableAsync(() =>
        fetch(`${this._serviceUrl}/${id}/${completed}`, {
          method: 'PATCH',
        })
      );

      if (res.status === 204) {
        return Promise.resolve();
      }

      return Promise.reject(`${res.status}: ${res.statusText}`);
    } catch (error) {
      return Promise.reject(`Fetch Error: ${error}`);
    }
  }

  async remove(id) {
    try {
      const res = await retryableAsync(() =>
        fetch(`${this._serviceUrl}/${id}`, {
          method: 'DELETE',
        })
      );

      if (res.status === 204) {
        return Promise.resolve();
      }

      return Promise.reject(`${res.status}: ${res.statusText}`);
    } catch (error) {
      return Promise.reject(`Fetch Error: ${error}`);
    }
  }
}

File: mobile_client/lib/services/todo_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;

import '../utils/observable_utils.dart';
import '../models/todo_items.dart';

abstract class TodoService {
  Future<TodoItems> getAll();
  Future<TodoItem> add(TodoItem item);
  Future<void> update(String id, TodoItem item);
  Future<void> updateCompleted(String id, bool completed);
  Future<void> remove(String id);
}

class TodoServiceImpl extends TodoService {
  final Uri serviceUri;

  TodoServiceImpl(this.serviceUri);

  Future<TodoItems> getAll() async {
    try {
      var res = await retryableAsync(() => http.get(this.serviceUri));

      if (res.statusCode == 200) {
        return TodoItems.fromJson(json.decode(res.body));
      }

      return Future.error('${res.statusCode}: ${res.body}');
    } catch (e) {
      return Future.error('Fetch Error: $e');
    }
  }

  Future<TodoItem> add(TodoItem item) async {
    try {
      var res = await retryableAsync(
        () => http.post(
              this.serviceUri,
              headers: {
                'Content-Type': 'application/json',
              },
              // body: JsonEncoder().convert(item.toJson()),
              body: item.toString(),
            ),
      );

      if (res.statusCode == 201) {
        return TodoItem.fromJson(json.decode(res.body));
      }

      return Future.error('${res.statusCode}: ${res.body}');
    } catch (e) {
      return Future.error('Fetch Error: $e');
    }
  }

  Future<void> update(String id, TodoItem item) async {
    try {
      var res = await retryableAsync(
        () => http.put(
              '${this.serviceUri}/$id',
              headers: {
                'Content-Type': 'application/json',
              },
              body: JsonEncoder().convert(item.toJson()),
            ),
      );

      if (res.statusCode == 204) {
        return Future.value();
      }

      return Future.error('${res.statusCode}: ${res.body}');
    } catch (e) {
      return Future.error('Fetch Error: $e');
    }
  }

  Future<void> updateCompleted(String id, bool completed) async {
    try {
      var res = await retryableAsync(
        () => http.patch('${this.serviceUri}/$id/$completed'),
      );

      if (res.statusCode == 204) {
        return Future.value();
      }

      return Future.error('${res.statusCode}: ${res.body}');
    } catch (e) {
      return Future.error('Fetch Error: $e');
    }
  }

  Future<void> remove(String id) async {
    try {
      var res = await retryableAsync(
        () => http.delete('${this.serviceUri}/$id'),
      );

      if (res.statusCode == 204) {
        return Future.value();
      }

      return Future.error('${res.statusCode}: ${res.body}');
    } catch (e) {
      return Future.error('Fetch Error: $e');
    }
  }
}
